applications/openbrush-follow-up-2.md

```diff
@@ -0,0 +1,369 @@
+|--------|----------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| 0a.    | License                                                                          | MIT                                                                                                                                                                                                                                                                                                                                                                                                   |
+| 0b.    | Documentation                                                                    | We will provide a documentation of how to write upgradable contracts                                                                                                                                                                                                                                                                                                                                  |
+| 1.     | Implement `delegate_call` in contract-pallet                                     | The ethereum has a [`delegatecall`](https://docs.soliditylang.org/en/v0.8.10/introduction-to-smart-contracts.html#delegatecall-callcode-and-libraries) function that helps to achieve upgradable functionality in contracts. We will add the support of that function into the `contract-pallet`. With that function the OpenBurhs will provide the implementation fo `Proxy` and `Diamond` patterns. |
+| 2.     | Implement `set_code_hash` in contract-pallet                                     | The usage of `Proxy` pattern adds overhead during each execution of the contract. It increase the cost of execution, and size of the proof of verification. Instead of `delegate_call` the `contract-pallet` can provide a separate function to change the source code fo the contract. It simplifies the flow and allow to easy upgrade the contract in the future.                                  |
+| 3.     | Import `delegate_call`, `set_code_hash` in ink! and update `Proxy` example       | Support of `delegate_call`, `set_code_hash` on ink! level and providing a user-friendly abstraction. The ink! already provides an abstraction for cross-contract communication, we will extend it to support `delegate_call`.                                                                                                                                                                         |
```

Comment:
Aren't those steps already done?

Reply:
Yes, it was done.

We created milestones 6-8 several months ago, and we've started work on the 6th milestone before publishing the Grant for it. 
We couldn't post it because the 4th milestone was not finished(we waited for the finalization of the PSPs). We've started work on the 5th milestone, but it depends on ink! team and we waited for feedback. At the same time, we started to participate in the Substrate Builders Program, where we represented milestones 6-8. We agreed that the Supercolony would not take money for the SBP, and we can apply with the same milestones for the Grant.

So we started work on it when:
- The milestone four was almost done we only waited for the finalization of PSPs.
- Part of milestone five was done, but it required feedback from ink! team.
- Milestone six is part of the SBP, but funds from the Grant that is not published.

We discussed this with the SBP team and they said that it is okay if a small part of the tasks will be ready in the grant.

We planned to publish it soon, but after:
- We got more feedback and changed the PSP, it took more time for finalization.
- We changed the scope of milestone 5 because refactoring of the storage is more critical and also affects part of milestone 6.

All that created delays to post the Grant 3. And it is why part of the tasks already are ready=)